modules/tinymce/src/plugins/directionality/test/ts/browser/DirectionalitySanityTest.ts

```diff
@@ -9,42 +8,189 @@ describe('browser.tinymce.plugins.directionality.DirectionalitySanityTest', () =
+    TinyAssertions.assertContent(editor, '<p dir="rtl">a</p>');
   });
 
   it('TBA: Set and select content, click on the Left to right toolbar button and assert direction is left to right', () => {
     const editor = hook.editor();
     editor.setContent('<p dir="rtl">a</p>');
     TinySelections.setSelection(editor, [ 0, 0 ], 0, [ 0, 0 ], 1);
     TinyUiActions.clickOnToolbar(editor, 'button[title="Left to right"]');
-    TinyAssertions.assertContentStructure(editor, ApproxStructure.build((s, str) => {
-      return s.element('body', {
-        children: [
-          s.element('p', {
-            attrs: {
-              dir: str.is('ltr')
-            }
-          })
-        ]
-      });
-    }));
+    TinyAssertions.assertContent(editor, '<p>a</p>'); // as the default dir is ltr it just removes the dir attr
+  });
+
+  it('TINY-4589: should set two paragraphs to rtl and ltl', () => {
+    const editor = hook.editor();
+    editor.setContent('<p>foo</p><p>bar</p>');
+    TinySelections.setSelection(editor, [ 0 ], 0, [ 1 ], 1);
+    TinyUiActions.clickOnToolbar(editor, 'button[title="Right to left"]');
+    TinyAssertions.assertContent(editor, '<p dir="rtl">foo</p><p dir="rtl">bar</p>');
+
+    TinySelections.setSelection(editor, [ 0 ], 0, [ 1 ], 1);
+    TinyUiActions.clickOnToolbar(editor, 'button[title="Left to right"]');
+    TinyAssertions.assertContent(editor, '<p>foo</p><p>bar</p>');
+  });
+
+  it('TINY-4589: should set parent dir when element is a list item', () => {
+    const editor = hook.editor();
+    editor.setContent('<ul><li>foo</li><li>bar</li></ul>');
+    TinySelections.setSelection(editor, [ 0, 0 ], 0, [ 0, 0 ], 1);
+    TinyUiActions.clickOnToolbar(editor, 'button[title="Right to left"]');
+    TinyAssertions.assertContent(editor, '<ul dir="rtl"><li>foo</li><li>bar</li></ul>');
+
+    TinyUiActions.clickOnToolbar(editor, 'button[title="Left to right"]');
+    TinyAssertions.assertContent(editor, '<ul><li>foo</li><li>bar</li></ul>');
+  });
+
+  it('TINY-4589: should remove dir from selected list item and children', () => {
+    const editor = hook.editor();
+    editor.setContent(
+      '<ul>' +
+        '<li dir="ltr">foo' +
+          '<ul>' +
+            '<li dir="ltr">a</li>' +
+            '<li dir="rtl">b</li>' +
+            '<li>c</li>' +
+          '</ul>' +
+        '</li>' +
+        '<li dir="xyz">bar</li>' +
+      '</ul>'
+    );
+    TinySelections.setSelection(editor, [ 0, 0 ], 0, [ 0, 0 ], 1);
+    TinyUiActions.clickOnToolbar(editor, 'button[title="Right to left"]');
+    TinyAssertions.assertContent(editor,
+      '<ul dir="rtl">' +
+        '<li>foo' +
+          '<ul>' +
+            '<li dir="ltr">a</li>' +
+            '<li dir="rtl">b</li>' +
+            '<li>c</li>' +
+          '</ul>' +
+        '</li>' +
+        '<li>bar</li>' +
+      '</ul>'
+    );
+  });
+
+  it('TINY-4589: applying the same dir makes no changes', () => {
+    const editor = hook.editor();
+    editor.setContent(
```

Comment:
This uses the same HTML twice, so can we pull this out into a separate variable please? That'll also guarantee it's 100% the same.

Reply:
Yeap, agree.